src/fcntl/statx.rs

```diff
@@ -0,0 +1,371 @@
+        STATX_MTIME;
+        /// Want stx_ctime
+        STATX_CTIME;
+        /// Want stx_ino
+        STATX_INO;
+        /// Want stx_size
+        STATX_SIZE;
+        /// Want stx_blocks
+        STATX_BLOCKS;
+        /// [All of the above]
+        STATX_BASIC_STATS;
+        /// Want stx_btime
+        STATX_BTIME;
+        /// Want stx_mnt_id (since Linux 5.8);
+        STATX_MNT_ID;
+        /// [All currently available fields]
+        STATX_ALL;
+    }
+);
+
+/*
+libc_bitflags!(
+    pub struct FileAttributes: c_int {
+        /// The file is compressed by the filesystem and may take extra resources to access.
+        STATX_ATTR_COMPRESSED;
+
+        /// The file cannot be modified: it cannot be deleted or renamed, no hard links can be created to this file and no data can be written to it.
+        STATX_ATTR_IMMUTABLE;
+
+        /// The file can only be opened in append mode for writing.  Random access writing is not permitted.
+        STATX_ATTR_APPEND;
+
+
+        /// File is not a candidate for backup when a backup program such as `dump` is run.
+        STATX_ATTR_NODUMP;
+
+        /// A key is required for the file to be encrypted by the filesystem.
+        STATX_ATTR_ENCRYPTED;
+
+        // ///  (since Linux 5.5)
+        // /// The  file  has fs-verity enabled.  It cannot be written to, and all reads from it will be verified
+        // /// against a cryptographic hash that covers the entire file
+        // /// (e.g., via a Merkle tree).
+        // STATX_ATTR_VERITY;
+
+        // /// (since Linux 5.8)
+        // /// The file is in the DAX (cpu direct access) state.
+        // STATX_ATTR_DAX;
+    }
+);
+*/
+
+/// Attempt to retrieve stats of `pathname`. If `pathname` is relative, `dirfd` is used as starting directory for lookups. If `dirfs` is None, current directory is used.
+/// `mask` determines what stats entries should be filled in, but kernel can actually fill more or less than requested.
+/// `statx` is not atomic: if attributes are being changed at the time of `statx` is called, the returned attributes set may have items from different moments of time.
+pub fn statx<P: ?Sized + NixPath>(
+    dirfd: Option<RawFd>,
+    pathname: &P,
+    flags: Flags,
+    mask: Mask,
+) -> Result<Statx> {
+    let dirfd = dirfd.unwrap_or(libc::AT_FDCWD);
+
+    let mut dst = mem::MaybeUninit::uninit();
+    let res = pathname.with_nix_path(|cstr| unsafe {
+        libc::statx(
+            dirfd,
+            cstr.as_ptr(),
+            flags.bits() as libc::c_int,
+            mask.bits() as libc::c_uint,
+            dst.as_mut_ptr(),
+        )
+    })?;
+
+    Errno::result(res)?;
+
+    Ok(Statx {
+        inner: unsafe { dst.assume_init() },
+    })
+}
+
+/// Get stats of specified file descriptor (not nesessarily directory) without resolving the path.
+/// Automaitcally sets the `AT_EMPTY_PATH` flag.
+pub fn statx_without_path(dirfd: RawFd, mut flags: Flags, mask: Mask) -> Result<Statx> {
```

Comment:
Is this function really necessary?  It seems like the user can accomplish the same thing with 
```rust
statx(Some(fd), "", Flags::AT_EMPTY_path, MASK)
```

Reply:
Yes, it is a helper function to call more generic `statx`. The syscall does different things based on parameters, so I expect it to be more readable and obvious if multiple functions represented different cases. For example, it is easy for forget about `Flags::AT_EMPTY_PATH` or wonder what to put inside the path argument, forcing user to stop coding and start looking at docs.
Also I expect even empty path should be terminated be zero byte, so just `""` may allocate (unless `with_nix_path` have some special optimisation).

Reply:
I'm leery of providing convenience methods like this.  Historically, whenever Nix goes beyond the minimum of "provide Rusty bindings" and invent a convenience API, we get it wrong.  Absent a more compelling reason, I would prefer to remove this function.